File: 910-Smallest Range II.py
class Solution(object):
    def smallestRangeII(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
# O(NlogN)
# https://jimmy-shen.medium.com/an-interesting-problem-910-smallest-range-ii-302ca119bc8c
        nums.sort()
        n = len(nums)
        res = nums[n-1] - nums[0]
        
        for i in range(n-1):
            mx = max(nums[i]+k, nums[n-1]-k)
            mn = min(nums[0]+k, nums[i+1]-k)
            res = min(res, mx-mn)
            
        return res

File: JavaScript/910-Smallest Range II.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var smallestRangeII = function(nums, k) {
    nums.sort(function(a,b){return a-b;})
    const N = nums.length;
    let res = nums[N-1] - nums[0];
    
    for (let i=0; i<N-1; i++){
        let mx = Math.max(nums[i]+k, nums[N-1]-k);
        let mn = Math.min(nums[0]+k, nums[i+1]-k);
        res = Math.min(res, mx-mn);
    }
    return res;
};